# Project GEP00001


```
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00001
# move previous analysis in 20171219_AmpliconSeqAnalysis/
```

## Align

```
scp shell/ngs/* clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00001/.

scp clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00001/samplesheet.csv .
scp clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00001/filelist.csv .
python python/scripts/convert_filelist.py > ~/mnt/scratchb/genome-editing/GEP00001/filelist.csv
python python/scripts/convert_samplesheet.py  > ~/mnt/scratchb/genome-editing/GEP00001/samplesheet.csv
```

```
vi alignment-meta.xml  # update to the latest version
sbatch job_alignment.sh
tail -f alignment.out
```

## Combine reads

```
sbatch job_joinreads.sh
tail -f joinreads.out
```

## Read counts

```
python python/scripts/create_pipeline_files.py --project=GEP00001 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa --seq-dict=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=/Users/pajon01/mnt/scratchb/genome-editing/GEP00001/filelist.csv

scp targets.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00001/.
scp amplicons.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00001/.
scp samples.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00001/.

scp amplicount_config.csv clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00001/.
```

```
sbatch job_amplicount.sh
tail -f amplicount.out

source /home/pajon01/genome-editing/venv/bin/activate
python /home/pajon01/genome-editing/python/scripts/ampliplot.py
```

## MultiQC report

```
sbatch job_fastqc.sh
tail -f fastqc.out

sbatch job_multiqc.sh
tail -f multiqc.out
```

## Run Amplicon sequencing pipeline

```
./configure_amplicon.sh GEP00001
sbatch job_amplicon_gatk.sh
tail -f amplicon-gatk.out

sbatch job_amplicon_vardict.sh
tail -f amplicon-vardict.out
```

## Amplican

Config file generator:
```
python python/scripts/create_amplican_conf.py --project=GEP00001 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa
```
