# Project GEP00013 (Liz)

target:
genome: Homo sapiens [GRCh38]
name: PIP4K2C
gene_id: ENST00000540759
chrom: 12
start: 57591203
end: 57603418
fprimer_seq: 'GGGCCCCAGAAATTCAAAGG'
rprimer_seq: 'CTGACTGGACCTCCTGCTCT'

http://qiime.org/scripts/join_paired_ends.html

```
# install fastq-join to join paired end fastq files on cluster home directory clust1-headnode /home/pajon01/fastq-join
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00013/fastq
for f in *.s_1.r_1.fq.gz; do ~/fastq-join/fastq-join $f `echo $f | cut -d'.' -f1-4`.r_2.fq.gz -o `echo $f | cut -d'.' -f1-4`.fq; done
for f in *.fqjoin; do gzip $f; done
```

```
# locally after mounting mntclustsb to get sequencing data
ll /Users/pajon01/mnt/scratchb/genome-editing/GEP00013
cd /Users/pajon01/workspace/genome-editing

# count the reads
python python/scripts/count_amplicons.py > GEP00013_counts_join.csv

# pairwise alignment
python python/scripts/align_amplicons.py GEP00013_counts_join.csv
```

```
#faToTwoBit in.fa out.2bit
cd /Users/pajon01/workspace/genome-editing/blat
./gfServer start localhost 8888 GRCh38.2bit -log=gfServer.log -canStop -stepSize=5 > gfServer.out &
./gfServer status localhost 8888

./gfServer pcr localhost 8888 GGGCCCCAGAAATTCAAAGG CTGACTGGACCTCCTGCTCT 500
GRCh38.2bit:X	122801770	122802180	+
GRCh38.2bit:1	26225280	26225715	-
GRCh38.2bit:22	25104385	25104655	-

./gfServer pcr localhost 8888 GGGCCCCAGAAATTCAAAGG AGAGCAGGAGGTCCAGTCAG 500
GRCh38.2bit:9	135971690	135972030	+
GRCh38.2bit:12	57600720	57600945	+
GRCh38.2bit:8	25155195	25155575	-

./gfServer pcr localhost 8888 GGGCCCCAGAAATTCAAAGG AGAGCAGGAGGTCCAGTCAG 300
GRCh38.2bit:12	57600720	57600945	+

./gfServer pcrDirect GGGCCCCAGAAATTCAAAGG AGAGCAGGAGGTCCAGTCAG GRCh38.2bit
Counting tiles in GRCh38.2bit
Done adding
Index built in 92.475 seconds
plus strand:
Index searched in 0.005 seconds
0-0 GRCh38.2bit:2 217733153-217735232, hits 0
0-0 GRCh38.2bit:2 91537820-91541593, hits 0
0-0 GRCh38.2bit:16 58263315-58265471, hits 0
minus strand:
Index searched in 0.002 seconds
0-0 GRCh38.2bit:4 186450000-186451331, hits 0
0-0 GRCh38.2bit:14 33952743-33954558, hits 0

./gfServer stop localhost 8888
```

## Joining reads
```
cp shell/ngs/job_joinreads.sh /Users/pajon01/mnt/scratchb/genome-editing/GEP00013/.
```

```
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00013/
sbatch job_joinreads.sh
```

## Configuring analysis

```
python python/dnascissors/utils.py GEP00013
GEP00013
Ensembl Gene ID	ENST00000540759
forward primer	GGGCCCCAGAAATTCAAAGG 9518
reverse primer	CTGACTGGACCTCCTGCTCT 9737
target seq	TACAGGTACAGGGGTGATACCTAGCAGTGAAGAGTGAGAGAGAGGTGTCTGATTTGTCAGTGGGTCTCAGTTTCTAGTGCAGCTGAAGATCATGGACTACAGCCTTCTGCTAGGCATCCACGACATCATTCGGGGCTCTGAACCAGAGGAGGAAGCGCCCGTGCGGGAGGATGAGTCAG
target seq len	179
amplicon coord	chr12	57600721	57600959	+	GRCh38_chr12_57600721
target coord	chr12	57600741	57600939	+	GRCh38_chr12_57600741
target len	199
target name	PIP4K2C
Target coordinates not equal. Do run blat!
```

```
./gfServer pcr localhost 8888 GGGCCCCAGAAATTCAAAGG AGAGCAGGAGGTCCAGTCAG 300
GRCh38_hs38d1.2bit:chr12	57600720	57600945	+

AGGGCCCCAGAAATTCAAAGG
TACAGGTACAGGGGTGATACCTAGCAGTGAAGAGTGAGAGAGAGGTGTCTGATTTGTCAGTGGGTCTCAGTTTCTAGTGCAGCTGAAGATCATGGACTACAGCCTTCTGCTAGGCATCCACGACATCATTCGGGGCTCTGAACCAGAGGAGGAAGCGCCCGTGCGGGAGGATGAGTCAG AGGTGGATGGGGACTGCAGCCTGACT
```

methodology:
- get amplicon start position 57600721 - 500 = 57600221
- get amplicon end position 57600721 + 500 = 57601221
- chr12:57600221-57601221
- `/home/bioinformatics/software/samtools/samtools-1.8/bin/samtools faidx /scratcha/bioinformatics/reference_data/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa chr12:57600221-57601221`
```
>chr12:57600221-57601221
GATGAGATCTAACTGAGCTAGCATAGAGCCCTAGACAGAAGGTGGGGTTCTGAGACCTTGGGGAAGGGATATGTTCCCAAGATGTCCCTTTACATATTCTTAGGTTAAAGAATTGCCCACCCTTAAGGATATGGACTTTCTCAACAAGAACCAGAAAGTATATATTGGTGAAGAGGAGAAGAAAATATTTCTGGAGAAGCTGAAGAGAGATGTGGAGGTGATGATTGGGTGCTCTGGGAAATGGCTTTCCTTTTTTTGCTCCCTAGAGGGTAGTTGTCTCTTTCATTCACGGTATGAGGGAGCTCCTCCCCTTCCCTTTATTCTTCAGGTCCTCTGCCTATATGGGGGTTTAGTATGAGACATGAAAAACCCCTGGAGCTGACCCGTGGAAGGGGTGGCTGGAATAAGCTAAAAGGATGATCTTAGATACTAGGACATATAGGCACTAAAGGCAGTGGTATGCCCTGCAACCTCAATTTTTATACTTCCTTTCCCCAGAA

GGGCCCCAGAAATTCAAAGG

TACAGGTACAGGGGTGATACCTAGCAGTGAAGAGTGAGAGAGAGGTGTCTGATTTGTCAGTGGGTCTCAGTTTCTAGTGCAGCTGAAGATCATGGACTACAGCCTTCTGCTAGGCATCCACGACATCATTCGGGGCTCTGAACCAGAGGAGGAAGCGCCCGTGCGGGAGGATGAGTCAGAGGTGGATGGGGACTGCAGC

CTGACTGGACCTCCTGCTCT

GGTGGGCTCCTATGGCACCTCCCCAGAGGGTATCGGAGGCTACATCCATTCCCATCGGCCCCTGGGCCCAGGAGAGTTTGAGTCCTTCATTGATGTCTATGCCATCCGGAGTGCTGAAGGTGAGAGAACCGGGACAATTTAAGGGTGGAGAGGGGATTTTTCCTGGAGGACAGAGACCAGAGTGCTGGGGGAGAGCCTGATTAGCTTTTCAGAACAAAACTGACTGCTTCCTGGAGAAAGGGAATTGGAACTAAACCTACTC
```
- find primer pairs in sequence to get coordinates
  - start amplicon: 57600221 + 500 = 57600721
  - end amplicon:   57601221 - 262 = 57600959
- test if coordinates are incorrect
`/home/bioinformatics/software/samtools/samtools-1.8/bin/samtools faidx /scratcha/bioinformatics/reference_data/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa chr12:57600721-57600959`
```
>chr12:57600721-57600959
GGGCCCCAGAAATTCAAAGGTACAGGTACAGGGGTGATACCTAGCAGTGAAGAGTGAGAG
AGAGGTGTCTGATTTGTCAGTGGGTCTCAGTTTCTAGTGCAGCTGAAGATCATGGACTAC
AGCCTTCTGCTAGGCATCCACGACATCATTCGGGGCTCTGAACCAGAGGAGGAAGCGCCC
GTGCGGGAGGATGAGTCAGAGGTGGATGGGGACTGCAGCCTGACTGGACCTCCTGCTCT
```
- final amplicon coordinates validated: chr12:57600721-57600959

### Testing pyfaidx

https://github.com/mdshw5/pyfaidx

```
sshfs pajon01@clust1-headnode.cri.camres.org:/mnt/scratcha/bioinformatics/reference_data/ /Users/pajon01/mnt/refdata
ll /Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa
cd /Users/pajon01/workspace/genome-editing/
source venv/bin/activate
pip install pyfaidx
python
>>> from pyfaidx import Fasta
>>> Fasta('/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa')['chr12'][57600221:57601221]
>chr12:57600222-57601221
ATGAGATCTAACTGAGCTAGCATAGAGCCCTAGACAGAAGGTGGGGTTCTGAGACCTTGGGGAAGGGATATGTTCCCAAGATGTCCCTTTACATATTCTTAGGTTAAAGAATTGCCCACCCTTAAGGATATGGACTTTCTCAACAAGAACCAGAAAGTATATATTGGTGAAGAGGAGAAGAAAATATTTCTGGAGAAGCTGAAGAGAGATGTGGAGGTGATGATTGGGTGCTCTGGGAAATGGCTTTCCTTTTTTTGCTCCCTAGAGGGTAGTTGTCTCTTTCATTCACGGTATGAGGGAGCTCCTCCCCTTCCCTTTATTCTTCAGGTCCTCTGCCTATATGGGGGTTTAGTATGAGACATGAAAAACCCCTGGAGCTGACCCGTGGAAGGGGTGGCTGGAATAAGCTAAAAGGATGATCTTAGATACTAGGACATATAGGCACTAAAGGCAGTGGTATGCCCTGCAACCTCAATTTTTATACTTCCTTTCCCCAGAAGGGCCCCAGAAATTCAAAGGTACAGGTACAGGGGTGATACCTAGCAGTGAAGAGTGAGAGAGAGGTGTCTGATTTGTCAGTGGGTCTCAGTTTCTAGTGCAGCTGAAGATCATGGACTACAGCCTTCTGCTAGGCATCCACGACATCATTCGGGGCTCTGAACCAGAGGAGGAAGCGCCCGTGCGGGAGGATGAGTCAGAGGTGGATGGGGACTGCAGCCTGACTGGACCTCCTGCTCTGGTGGGCTCCTATGGCACCTCCCCAGAGGGTATCGGAGGCTACATCCATTCCCATCGGCCCCTGGGCCCAGGAGAGTTTGAGTCCTTCATTGATGTCTATGCCATCCGGAGTGCTGAAGGTGAGAGAACCGGGACAATTTAAGGGTGGAGAGGGGATTTTTCCTGGAGGACAGAGACCAGAGTGCTGGGGGAGAGCCTGATTAGCTTTTCAGAACAAAACTGACTGCTTCCTGGAGAAAGGGAATTGGAACTAAACCTACTC
```

```
python python/scripts/amplifind.py GEP00013 /Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa
Project ID	GEP00013
/Users/pajon01/workspace/genome-editing/venv/lib/python3.6/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
--- Amplicon #1
Target name	PIP4K2C
Forward primer	GGGCCCCAGAAATTCAAAGG
Reverse primer	CTGACTGGACCTCCTGCTCT
target coord	chr12	57600741	57600939	+	chr12_57600721
amplicon coord	chr12	57600721	57600959	+	chr12_57600721
amplicon desc	chr12:57600721:57600959:+
amplicon seq	GGGCCCCAGAAATTCAAAGGTACAGGTACAGGGGTGATACCTAGCAGTGAAGAGTGAGAGAGAGGTGTCTGATTTGTCAGTGGGTCTCAGTTTCTAGTGCAGCTGAAGATCATGGACTACAGCCTTCTGCTAGGCATCCACGACATCATTCGGGGCTCTGAACCAGAGGAGGAAGCGCCCGTGCGGGAGGATGAGTCAGAGGTGGATGGGGACTGCAGCCTGACTGGACCTCCTGCTCT
amplicon seq len	239
---
```


## Amplicon pipeline

```
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00013/
./configure_amplicon.sh GEP00013
source ~/genome-editing/venv/bin/activate
python ~/genome-editing/python/scripts/create_pipeline_files.py --project=GEP00013 --seq-dict=/scratcha/bioinformatics/reference_data/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=filelist.csv
Amplicon 51 12 57600721-57600959 +
Has primer 99 57600721-57600740 forward GGGCCCCAGAAATTCAAAGG
Has primer 100 57600940-57600959 reverse CTGACTGGACCTCCTGCTCT
tail -f amplicons.txt
chr12	57600721	57600959	+	GRCh38_chr12_57600721
tail -f targets.txt
chr12	57600741	57600939	+	GRCh38_chr12_57600721
sbatch job_amplicon_vardict.sh
sbatch job_amplicon_gatk.sh
```

## Load results

```
shell/load_project_GEP00013.sh
```

## Read counts

```
python python/scripts/create_pipeline_files.py --project=GEP00013 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa --seq-dict=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=/Users/pajon01/mnt/scratchb/genome-editing/GEP00013/filelist.csv

scp amplicount_config.csv clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00013/.
scp shell/ngs/job_amplicount.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00013/.
```

```
sbatch job_amplicount.sh
tail -f amplicount.out

source /home/pajon01/genome-editing/venv/bin/activate
python /home/pajon01/genome-editing/python/scripts/ampliplot.py
```

## Publish

```
scp shell/ngs/job_publish.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00013/.
sbatch job_publish.sh GEP00013
```

## Re-load growth data (Incucyte)

```
PROJNAME="GEP00013"
PROJDIR=data/${PROJNAME}
echo 'Loading project' ${PROJNAME}
for plateid in 01 02; do echo 'Loading data for plate' ${plateid}; python python/scripts/load_cell_growth.py --plateid=${PROJNAME}_${plateid} --file=${PROJDIR}/${PROJNAME}_${plateid}_1in5_incu.txt --clean; done
```
