# Project GEP00013 (Liz)

target:
genome: Homo sapiens [GRCh38]
name: PIP4K2C
gene_id: ENST00000540759
chrom: 12
start: 57591203
end: 57603418
fprimer_seq: 'GGGCCCCAGAAATTCAAAGG'
rprimer_seq: 'CTGACTGGACCTCCTGCTCT'

http://qiime.org/scripts/join_paired_ends.html

```
# install fastq-join to join paired end fastq files on cluster home directory clust1-headnode /home/pajon01/fastq-join
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00013/fastq
for f in *.s_1.r_1.fq.gz; do ~/fastq-join/fastq-join $f `echo $f | cut -d'.' -f1-4`.r_2.fq.gz -o `echo $f | cut -d'.' -f1-4`.fq; done
for f in *.fqjoin; do gzip $f; done
```

```
# locally after mounting mntclustsb to get sequencing data
ll /Users/pajon01/mnt/scratchb/genome-editing/GEP00013
cd /Users/pajon01/workspace/genome-editing

# count the reads
python python/scripts/count_amplicons.py > GEP00013_counts_join.csv

# pairwise alignment
python python/scripts/align_amplicons.py GEP00013_counts_join.csv
```

```
#faToTwoBit in.fa out.2bit
cd /Users/pajon01/workspace/genome-editing/blat
./gfServer start localhost 8888 GRCh38.2bit -log=gfServer.log -canStop -stepSize=5 > gfServer.out &
./gfServer status localhost 8888

./gfServer pcr localhost 8888 GGGCCCCAGAAATTCAAAGG CTGACTGGACCTCCTGCTCT 500
GRCh38.2bit:X	122801770	122802180	+
GRCh38.2bit:1	26225280	26225715	-
GRCh38.2bit:22	25104385	25104655	-

./gfServer pcr localhost 8888 GGGCCCCAGAAATTCAAAGG AGAGCAGGAGGTCCAGTCAG 500
GRCh38.2bit:9	135971690	135972030	+
GRCh38.2bit:12	57600720	57600945	+
GRCh38.2bit:8	25155195	25155575	-

./gfServer pcr localhost 8888 GGGCCCCAGAAATTCAAAGG AGAGCAGGAGGTCCAGTCAG 300
GRCh38.2bit:12	57600720	57600945	+

./gfServer stop localhost 8888
```

## Joining reads
```
cp shell/ngs/job_joinreads.sh /Users/pajon01/mnt/scratchb/genome-editing/GEP00013/.
```

```
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00013/
sbatch job_joinreads.sh
```

## Configuring analysis

```
python python/dnascissors/utils.py GEP00013
GEP00013
Ensembl Gene ID	ENST00000540759
forward primer	GGGCCCCAGAAATTCAAAGG 9518
reverse primer	CTGACTGGACCTCCTGCTCT 9737
target seq	TACAGGTACAGGGGTGATACCTAGCAGTGAAGAGTGAGAGAGAGGTGTCTGATTTGTCAGTGGGTCTCAGTTTCTAGTGCAGCTGAAGATCATGGACTACAGCCTTCTGCTAGGCATCCACGACATCATTCGGGGCTCTGAACCAGAGGAGGAAGCGCCCGTGCGGGAGGATGAGTCAG
target seq len	179
amplicon coord	chr12	57600721	57600959	+	GRCh38_chr12_57600721
target coord	chr12	57600741	57600939	+	GRCh38_chr12_57600741
target len	199
target name	PIP4K2C
Target coordinates not equal. Do run blat!
```
