# Project GEP00009

## Sequencing data
- SLX-15026
- MiSeq PE300
- 180112_M01686_0235_000000000-BJWVR

## Combining paired reads
Target lengths are smaller than each read lengths, so we will combine R1 and R2 into one file

```
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00009/fastq

# reverse complement R2
for f in *.s_1.r_2.fq.gz; do seqkit seq -r -p $f -o `echo $f | cut -d'.' -f1-5`.fqrc.gz; done

# combine R1 & R2
for f in *.s_1.r_1.fq.gz; do zcat $f `echo $f | cut -d'.' -f1-4`.r_2.fqrc.gz | gzip > `echo $f | cut -d'.' -f1-4`.fqjoin.gz; done
```

### FastQC & MultiQC

```
scp shell/ngs/job_fastqc.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00009/.
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00009/
sbatch job_fastqc.sh
```

```
scp shell/ngs/job_multiqc.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00009/.
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00009/
sbatch job_multiqc.sh
```

## Amplicon pipeline

- Amplicon / target coordinates

```
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00009
tail -f targets.txt
chr16	53704149	53704290	+	GRCh38_chr16_53704130
chr16	54931196	54931376	+	GRCh38_chr16_54931181
tail -f amplicons.txt
chr16	53704130	53704310	+	GRCh38_chr16_53704130
chr16	54931181	54931395	+	GRCh38_chr16_54931181
```

```
python python/scripts/amplifind.py GEP00009
```

- New reference genome, need to be re-align

```
/scratchb/bioinformatics/pajon01/genome-editing/GEP00009
# move previous analysis in 20180119_AmpliconSeqAnalysis/
```

```
scp shell/ngs/* clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00009/.
```

```
cp ~/mnt/scratchb/genome-editing/GEP00009/samplesheet.csv .
python python/scripts/convert_samplesheet.py > ~/mnt/scratchb/genome-editing/GEP00009/samplesheet.csv

cp ~/mnt/scratchb/genome-editing/GEP00009/filelist.csv .
python python/scripts/convert_filelist.py > ~/mnt/scratchb/genome-editing/GEP00009/filelist.csv
```

```
sbatch job_alignment.sh
```

- Re-run pipeline

```
python python/scripts/create_pipeline_files.py --project=GEP00009 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa --seq-dict=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=/Users/pajon01/mnt/scratchb/genome-editing/GEP00009/filelist.csv
scp targets.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00009/.
scp amplicons.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00009/.
scp samples.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00009/.
```

```
./configure_amplicon.sh GEP00009
sbatch job_amplicon_gatk.sh
sbatch job_amplicon_vardict.sh
```

re-run gatk pipeline with the same filtering as GEP00010 and GEP00005:
```
# <snvFilterExpression><![CDATA[QD < 0.1,FS > 60.0,MQ < 40.0,MQRankSum < -12.5]]></snvFilterExpression>
# <indelFilterExpression><![CDATA[QD < 0.1,FS > 200.0]]></indelFilterExpression>

vi config.gatk.xml
sbatch job_amplicon_gatk.sh
```

## Read counts

```
python python/scripts/create_pipeline_files.py --project=GEP00009 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa --seq-dict=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=/Users/pajon01/mnt/scratchb/genome-editing/GEP00009/filelist.csv

scp amplicount_config.csv clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00009/.
scp shell/ngs/job_amplicount.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00009/.

ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00009/
sbatch job_amplicount.sh
tail -f amplicount.out
```
