# Project GEP00005

## Sequencing data
- SLX-15021
- MiSeq PE300
- 171204_M01686_0227_000000000-BJ8JD

## Combining paired reads
Target lengths are smaller than each read lengths, so we will combine R1 and R2 into one file

```
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00005/fastq

# reverse complement R2
for f in *.s_1.r_2.fq.gz; do seqkit seq -r -p $f -o `echo $f | cut -d'.' -f1-5`.fqrc.gz; done

# combine R1 & R2
for f in *.s_1.r_1.fq.gz; do zcat $f `echo $f | cut -d'.' -f1-4`.r_2.fqrc.gz | gzip > `echo $f | cut -d'.' -f1-4`.fqjoin.gz; done
```

## Read counts

### Targets
- FTO
gene_id: 'ENSG00000140718'
fprimer_seq: 'TCCAGGGCGAGGGATCTAC'
rprimer_seq: 'GCCATCAAACCAAGCCCTTC'
guide_seq: 'GGAACGAGAGCGCGAAGCTAAGG'

- IRX5
gene_id: 'ENSG00000176842'
fprimer_seq: 'CCATGCCCGTGTGTG'
rprimer_seq: 'TGGAGGTGCGAGTTGTAGC'
guide_seq: 'CCTCGCTGGCGCTCTACTCGTGC'

- IRX3
gene_id: 'ENSG00000177508'
fprimer_seq: 'TTACCAGCTGCGGGAAGATG'
rprimer_seq: 'GGAGCCTCGGAGCTGAAC'
guide_seq: 'CCGTACACGGACGAGAGCACGTT'

- PREPL
gene_id: 'ENSG00000138078'
fprimer_seq: ''
rprimer_seq: ''
guide_seq: ''

- LEPRLF
gene_id: ''
fprimer_seq: ''
rprimer_seq: ''
guide_seq: ''

- LEPRY2
gene_id: ''
fprimer_seq: ''
rprimer_seq: ''
guide_seq: ''

- LEPRY3
gene_id: ''
fprimer_seq: ''
rprimer_seq: ''
guide_seq: ''

- LEPRY4
gene_id: ''
fprimer_seq: ''
rprimer_seq: ''
guide_seq: ''

- RPGRIP1L
gene_id: ''
fprimer_seq: ''
rprimer_seq: ''
guide_seq: ''


### Run scripts

```
cd /Users/pajon01/workspace/genome-editing
cp shell/ngs/job_readcounts.sh /Users/pajon01/mnt/scratchb/genome-editing/GEP00005v1/.
cp GEP00005_IRX5.yml /Users/pajon01/mnt/scratchb/genome-editing/GEP00005v1/.
cp GEP00005_FTO.yml /Users/pajon01/mnt/scratchb/genome-editing/GEP00005v1/.
```

```
ssh clust1-headnode
cp job_readcounts.sh job_readcounts_FTO.sh
cp job_readcounts.sh job_readcounts_IRX5.sh
rm job_readcounts.sh

vi GEP00005_FTO.yml
vi job_readcounts_FTO.sh
sbatch job_readcounts_FTO.sh

vi GEP00005_IRX5.yml
vi job_readcounts_IRX5.sh
sbatch job_readcounts_IRX5.sh

vi GEP00005_IRX3.yml
vi job_readcounts_IRX3.sh
sbatch job_readcounts_IRX3.sh
```

### FastQC & MultiQC

```
scp shell/ngs/job_fastqc.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00005v1/.
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00005v1/
sbatch job_fastqc.sh
```

```
scp shell/ngs/job_multiqc.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00005v1/.
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00005v1/
sbatch job_multiqc.sh
```

- Empty fastq files:
```
grep WARNING multiqc.out
[2019-03-01 19:04:56,252] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0195.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,292] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0199.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,417] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0208.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,503] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0212.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,512] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0214.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,514] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0215.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,586] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0219.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,596] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0220.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,605] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0222.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,738] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0231.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,795] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0235.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,805] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0237.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,830] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0239.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,872] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0243.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,880] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0245.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,881] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0246.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,921] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0248.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:56,961] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0251.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:57,000] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0255.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:57,002] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0256.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:57,231] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0271.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:57,300] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0277.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:57,301] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0278.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:57,303] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0279.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:57,358] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0283.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
[2019-03-01 19:04:57,418] multiqc.modules.fastqc.fastqc                      [WARNING]  Couldn't read 'SLX-15021.FLD0288.000000000-BJ8JD.s_1.fqjoin_fastqc.zip' - Bad zip file
```

## Amplicon pipeline

- Amplicon / target coordinates

```
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00005
tail -f targets.txt
chr1	65636127	65636321	+	GRCh38_chr1_65636106
chr1	65636437	65636536	+	GRCh38_chr1_65636416
chr1	65636656	65636844	+	GRCh38_chr1_65636635
chr1	65636867	65637081	+	GRCh38_chr1_65636846
chr16	53700644	53700845	+	GRCh38_chr16_53700624
chr16	53704149	53704290	+	GRCh38_chr16_53704130
chr16	54285800	54285918	+	GRCh38_chr16_54285780
chr16	54931196	54931376	+	GRCh38_chr16_54931181
chr2	44344499	44344584	+	GRCh38_chr2_44344474
tail -f amplicons.txt
chr1	65636106	65636342	+	GRCh38_chr1_65636106
chr1	65636416	65636557	+	GRCh38_chr1_65636416
chr1	65636635	65636865	+	GRCh38_chr1_65636635
chr1	65636846	65637102	+	GRCh38_chr1_65636846
chr16	53700624	53700865	+	GRCh38_chr16_53700624
chr16	53704130	53704310	+	GRCh38_chr16_53704130
chr16	54285780	54285936	+	GRCh38_chr16_54285780
chr16	54931181	54931395	+	GRCh38_chr16_54931181
chr2	44344474	44344605	+	GRCh38_chr2_44344474
```

```
python python/scripts/amplifind.py GEP00005
```

- New reference genome, need to be re-align

```
/scratchb/bioinformatics/pajon01/genome-editing/GEP00005v1
# move previous analysis in 20171219_AmpliconSeqAnalysis/
mv ../GEP00005/* 20180209_AmpliconSeqAnalysis/.
cd ..
rm -rf GEP00005
mv GEP00005v1 GEP00005
```

```
scp shell/ngs/* clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00005/.
```

```
cp ~/mnt/scratchb/genome-editing/GEP00005/samplesheet.csv .
python python/scripts/convert_samplesheet.py > ~/mnt/scratchb/genome-editing/GEP00005/samplesheet.csv

cp ~/mnt/scratchb/genome-editing/GEP00005/filelist.csv .
python python/scripts/convert_filelist.py > ~/mnt/scratchb/genome-editing/GEP00005/filelist.csv
```

```
sbatch job_alignment.sh
```

- Re-run pipeline

```
python python/scripts/create_pipeline_files.py --project=GEP00005 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa --seq-dict=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=/Users/pajon01/mnt/scratchb/genome-editing/GEP00005/filelist.csv
scp targets.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00005/.
scp amplicons.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00005/.
scp samples.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00005/.
```

```
./configure_amplicon.sh GEP00005
sbatch job_amplicon_gatk.sh
sbatch job_amplicon_vardict.sh
```
