# Project GEP00010

## Sequencing data
- SLX-15025
- MiSeq PE300
- 180320_M01712_0010_000000000-BNNFL

## Combining paired reads
Target lengths are smaller than each read lengths, so we will combine R1 and R2 into one file

```
scp shell/ngs/job_mergereads.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.

ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010/
sbatch job_mergereads.sh
tail -f mergereads.out
```

## FastQC & MultiQC

```
scp shell/ngs/job_fastqc.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010/
sbatch job_fastqc.sh
```

```
scp shell/ngs/job_multiqc.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010/
sbatch job_multiqc.sh
```

## Amplicon pipeline

- Amplicon / target coordinates

```
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010
tail -f targets.txt
chr1	65636884	65637076	-	GRCh38_chr1_65636864
chr16	53704149	53704290	+	GRCh38_chr16_53704130
chr4	26481557	26481765	-	GRCh38_chr4_26481537
tail -f amplicons.txt
chr1	65636864	65637096	-	GRCh38_chr1_65636864
chr16	53704130	53704310	+	GRCh38_chr16_53704130
chr4	26481537	26481784	-	GRCh38_chr4_26481537
```

```
python python/geneditidtools/amplifind.py GEP00010
```

- Re-run pipeline

```
/scratchb/bioinformatics/pajon01/genome-editing/GEP00010
# move previous analysis in 20180503_AmpliconSeqAnalysis/
```

```
scp shell/ngs/* clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
```

- New reference genome, need to be re-align

```
vi alignment-meta.xml
head -1 samplesheet.csv
"SampleName","SLXId","Barcode"  # should be "SampleName","FSFSampleName","SampleBAM"

head -1 filelist.csv
"SampleName","SourceID","Filename","Flowcell","Lane","SequencingPlatform","PlatformModel","SequencingCentre","SequencingDate","EndType","ReadLength"
# "SampleName","SLXId","Barcode","Filename","Flowcell","Lane","SequencingPlatform","PlatformModel","SequencingCentre","SequencingDate","EndType","ReadLength"
```

```
cp ~/mnt/scratchb/genome-editing/GEP00010/samplesheet.csv .
python python/geneditidtools/convert_samplesheet.py > ~/mnt/scratchb/genome-editing/GEP00010/samplesheet.csv

cp ~/mnt/scratchb/genome-editing/GEP00010/filelist.csv .
python python/geneditidtools/convert_filelist.py > ~/mnt/scratchb/genome-editing/GEP00010/filelist.csv
```

```
sbatch job_alignment.sh
```

```
python python/geneditidtools/create_pipeline_files.py --project=GEP00010 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa --seq-dict=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=/Users/pajon01/mnt/scratchb/genome-editing/GEP00010/filelist.csv
scp targets.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
scp amplicons.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
scp samples.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
```

```
./configure_amplicon.sh GEP00010
sbatch job_amplicon_gatk.sh
sbatch job_amplicon_vardict.sh
```

## Read counts

```
# locally after mounting cluster reference and project directories
sshfs pajon01@clust1-headnode.cri.camres.org:/mnt/scratcha/bioinformatics/reference_data/ /Users/pajon01/mnt/refdata
sshfs pajon01@clust1-headnode.cri.camres.org:/mnt/scratchb/bioinformatics/pajon01 /Users/pajon01/mnt/scratchb

ll /Users/pajon01/mnt/scratchb/genome-editing/GEP00010
cd /Users/pajon01/workspace/genome-editing
python python/geneditidtools/amplicount.py --fastqdir=/Users/pajon01/mnt/scratchb/genome-editing/GEP00010/fastq/

# config file generated with python/geneditidtools/create_pipeline_files.py
scp amplicount_config.csv clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
scp shell/ngs/job_amplicount.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.

# run on cluster
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010/
sbatch job_amplicount.sh
tail -f amplicount.out

source /home/pajon01/genome-editing/venv/bin/activate
python /home/pajon01/genome-editing/python/geneditidtools/ampliplot.py

# umount
umount /Users/pajon01/mnt/scratchb
umount /Users/pajon01/mnt/refdata
```

## Publish data

```
sbatch job_publish.sh GEP00010
tail -f publish.out
```

## Read counts - consequence categorisation

```
scp clust1-headnode.cri.camres.org:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/amplicount_coverage.csv .
scp clust1-headnode.cri.camres.org:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/amplicount.csv .
scp clust1-headnode.cri.camres.org:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/amplicount_config.csv .

python python/geneditidtools/ampliplot.py
```

## Read counts - rare sequence analysis

```
[pajon01@clust1-headnode GEP00010]$ for i in `seq 193 196`; do mv fastq_all/SLX-15025.FLD0$i.000000000-BNNFL.s_1.fqjoin.gz fastq/.; done
[pajon01@clust1-headnode GEP00010]$ for i in `seq 205 208`; do mv fastq_all/SLX-15025.FLD0$i.000000000-BNNFL.s_1.fqjoin.gz fastq/.; done
[pajon01@clust1-headnode GEP00010]$ for i in `seq 217 220`; do mv fastq_all/SLX-15025.FLD0$i.000000000-BNNFL.s_1.fqjoin.gz fastq/.; done
[pajon01@clust1-headnode GEP00010]$ for i in `seq 329 332`; do mv fastq_all/SLX-15025.FLD0$i.000000000-BNNFL.s_1.fqjoin.gz fastq/.; done
[pajon01@clust1-headnode GEP00010]$ for i in `seq 341 344`; do mv fastq_all/SLX-15025.FLD0$i.000000000-BNNFL.s_1.fqjoin.gz fastq/.; done
[pajon01@clust1-headnode GEP00010]$ for i in `seq 353 356`; do mv fastq_all/SLX-15025.FLD0$i.000000000-BNNFL.s_1.fqjoin.gz fastq/.; done

```
