# Project GEP00010

## Sequencing data
- SLX-15025
- MiSeq PE300
- 180320_M01712_0010_000000000-BNNFL

## Combining paired reads
Target lengths are smaller than each read lengths, so we will combine R1 and R2 into one file

```
scp shell/ngs/job_mergereads.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.

ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010/
sbatch job_mergereads.sh
tail -f mergereads.out
```

## Read counts

### Targets

- NC_000016.10 (FTO) ENSG00000140718
gene_id: 'ENSG00000140718'
fprimer_seq: 'TCCAGGGCGAGGGATCTAC'
rprimer_seq: 'GCCATCAAACCAAGCCCTTC'
guide_seq: 'GGAACGAGAGCGCGAAGCTAAGG'

- NC_000001.11 (LEPR) ENSG00000116678
gene_id: 'ENSG00000116678'
fprimer_seq: 'TCCAGGACAGTTGCTCACAC'
rprimer_seq: 'TCTCTCTCCCACCCACAACT'
guide_seq: 'CCTAACTGTGTAATTTCACTGAA'

- NC_000004.12 (CCKAR) ENSG00000163394
gene_id: 'ENSG00000163394'
fprimer_seq: 'TTCATGGCCACCTTCCCCT'
rprimer_seq: 'TCTCCATCAGCTCTGCTCCT'
guide_seq: 'CCAGTGAGATGTCCCCTGACCCT'

### Run scripts
```
# locally after mounting mntclustsb to get sequencing data
mntclustsb
ll /Users/pajon01/mnt/scratchb/genome-editing/GEP00010
cd /Users/pajon01/workspace/genome-editing

# configure
vi GEP00010_CCKAR.yml
vi GEP00010_LEPR.yml
vi GEP00010_FTO.yml

# count the reads
python python/scripts/count_amplicons.py GEP00010_CCKAR.yml > GEP00010_counts_CCKAR_join.csv
python python/scripts/count_amplicons.py GEP00010_LEPR.yml > GEP00010_counts_LEPR_join.csv
python python/scripts/count_amplicons.py GEP00010_FTO.yml > GEP00010_counts_FTO_join.csv

python python/scripts/align_amplicons.py GEP00010_counts_CCKAR_join.csv CCAGTGAGATGTCCCCTGACCCT
python python/scripts/align_amplicons.py GEP00010_counts_LEPR_join.csv CCTAACTGTGTAATTTCACTGAA
python python/scripts/align_amplicons.py GEP00010_counts_FTO_join.csv GGAACGAGAGCGCGAAGCTAAGG

umount /Users/pajon01/mnt/scratchb
```

### FastQC & MultiQC

```
scp shell/ngs/job_fastqc.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010/
sbatch job_fastqc.sh
```

```
scp shell/ngs/job_multiqc.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010/
sbatch job_multiqc.sh
```

## Amplicon pipeline

- Amplicon / target coordinates

```
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00010
tail -f targets.txt
chr1	65636884	65637076	-	GRCh38_chr1_65636864
chr16	53704149	53704290	+	GRCh38_chr16_53704130
chr4	26481557	26481765	-	GRCh38_chr4_26481537
tail -f amplicons.txt
chr1	65636864	65637096	-	GRCh38_chr1_65636864
chr16	53704130	53704310	+	GRCh38_chr16_53704130
chr4	26481537	26481784	-	GRCh38_chr4_26481537
```

```
python python/scripts/amplifind.py GEP00010
```

- Re-run pipeline

```
/scratchb/bioinformatics/pajon01/genome-editing/GEP00010
# move previous analysis in 20180503_AmpliconSeqAnalysis/
```

```
scp shell/ngs/* clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
```

- New reference genome, need to be re-align

```
vi alignment-meta.xml
head -1 samplesheet.csv
"SampleName","SLXId","Barcode"  # should be "SampleName","FSFSampleName","SampleBAM"

head -1 filelist.csv
"SampleName","SourceID","Filename","Flowcell","Lane","SequencingPlatform","PlatformModel","SequencingCentre","SequencingDate","EndType","ReadLength"
# "SampleName","SLXId","Barcode","Filename","Flowcell","Lane","SequencingPlatform","PlatformModel","SequencingCentre","SequencingDate","EndType","ReadLength"
```

```
cp ~/mnt/scratchb/genome-editing/GEP00010/samplesheet.csv .
python python/scripts/convert_samplesheet.py > ~/mnt/scratchb/genome-editing/GEP00010/samplesheet.csv

cp ~/mnt/scratchb/genome-editing/GEP00010/filelist.csv .
python python/scripts/convert_filelist.py > ~/mnt/scratchb/genome-editing/GEP00010/filelist.csv
```

```
sbatch job_alignment.sh
```

```
python python/scripts/create_pipeline_files.py --project=GEP00010 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa --seq-dict=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=/Users/pajon01/mnt/scratchb/genome-editing/GEP00010/filelist.csv
scp targets.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
scp amplicons.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
scp samples.txt clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00010/.
```

```
./configure_amplicon.sh GEP00010
sbatch job_amplicon_gatk.sh
sbatch job_amplicon_vardict.sh
```
