# Project GEP00012 (Amy)

## Old coordinates:
targets
chr4	105233919	105234088	-	GRCh38_chr4_105233898
chr4	105233919	105234088	-	GRCh38_chr4_105233898

amplicons
chr4	105233898	105234109	-	GRCh38_chr4_105233898
chr4	105233898	105234109	-	GRCh38_chr4_105233898


## Re-ran pipeline with new coordinates
New amplicon coord based on searching primer sequence in gene
chr4    105146256       105146426       +       GRCh38_chr4_105145875
chr4	105146256	105146426	-	GRCh38_chr4_105145875

New target coord:
chr4    105145875       105146446       +       GRCh38_chr4_105145875
chr4	105145875	105146446	-	GRCh38_chr4_105145875

## Ran primer check using blat/gfServer query
nohup /home/bioinformatics/software/ucsc-tools/ucsc-tools-20171107/blat/gfServer start localhost 8888 GRCh38.2bit -log=gfServer.log -canStop -stepSize=5 > gfServer.out &
/home/bioinformatics/software/ucsc-tools/ucsc-tools-20171107/blat/gfServer pcr localhost 8888 TCAACTAGAGGGCAGCCTTG GGAGACACCAAGTGGCACTC 500
nohup /home/bioinformatics/software/ucsc-tools/ucsc-tools-20171107/blat/gfServer stop localhost 8888
No amplicon found on chromosome 4 !!!
Gene NM_017628 on chrom4 https://www.ncbi.nlm.nih.gov/nuccore/NM_017628.4
GRCh38.2bit:22	49994655	49994895	+
GRCh38.2bit:15	24117110	24117415	-

## Finding where the reads pile up
samtools view SLX-15101.FLD0045.000000000-D3PMP.s_1.bwamem.homo_sapiens.bam | cut -f3,4 | more
chr4	105233900

## Finding the region guessing the end of the region
samtools faidx /mnt/scratcha/bioinformatics/reference_data/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa chr4:105233900-105234109

>chr4:105233900-105234109
TCAACTAGAGGGCAGCCTTGTGGATGGCCCCGAAGCAAGCCTGATGGAACAGGATAGAAC
CAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCA
GACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCC
AGAAGTAAATGGAGACACCAAGTGGCACTC

TCAACTAGAGGGCAGCCTTG
GGAGACACCAAGTGGCACTC

target:
chr4	105233920	105234089	-	GRCh38_chr4_105233900

amplicon:
chr4	105233900	105234109	-	GRCh38_chr4_105233900

## Other ideas to check later
Open bam files in IGV
gatk locite

## Amplicon coordinate using blat

```
./gfServer pcrDirect TCAACTAGAGGGCAGCCTTG GGAGACACCAAGTGGCACTC GRCh38.2bit
Counting tiles in GRCh38.2bit
Done adding
Index built in 90.053 seconds
plus strand:
Index searched in 0.000 seconds
minus strand:
Index searched in 0.001 seconds
0-0 GRCh38.2bit:10 110035134-110037565, hits 0

./gfServer start localhost 8888 GRCh38.2bit -log=gfServer.log -canStop -stepSize=5 > gfServer.out &
./gfServer pcr localhost 8888 TCAACTAGAGGGCAGCCTTG GGAGACACCAAGTGGCACTC 500
GRCh38.2bit:22	49994655	49994895	+
GRCh38.2bit:15	24117110	24117415	-
```

## Merging reads

```
cp shell/ngs/job_mergereads.sh /Users/pajon01/mnt/scratchb/genome-editing/GEP00012/.
```

```
ssh clust1-headnode
cd /scratchb/bioinformatics/pajon01/genome-editing/GEP00012/
sbatch job_mergereads.sh
```


## Read counts

```
python python/scripts/create_pipeline_files.py --project=GEP00012 --genome=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.fa --seq-dict=/Users/pajon01/mnt/refdata/reference_genomes/homo_sapiens/GRCh38_hs38d1/fasta/hsa.GRCh38_hs38d1.dict --filelist=/Users/pajon01/mnt/scratchb/genome-editing/GEP00012/filelist.csv

scp amplicount_config.csv clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00012/.
scp shell/ngs/job_amplicount.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00012/.
```

```
sbatch job_amplicount.sh
tail -f amplicount.out

source /home/pajon01/genome-editing/venv/bin/activate
python /home/pajon01/genome-editing/python/scripts/ampliplot.py
```

## Publish

```
scp shell/ngs/job_publish.sh clust1-headnode:/scratchb/bioinformatics/pajon01/genome-editing/GEP00012/.
```

```
sbatch job_publish.sh GEP00012
```
